docs/examples/custom_channel.py

```diff
@@ -0,0 +1,131 @@
+#
+# Copyright (c) 2024 ZettaScale Technology
+#
+# This program and the accompanying materials are made available under the
+# terms of the Eclipse Public License 2.0 which is available at
+# http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+# which is available at https://www.apache.org/licenses/LICENSE-2.0.
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+#
+# Contributors:
+#   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+#
+import threading
+import time
+from typing import Callable
+
+import zenoh
+
+# Open session
+session = zenoh.open(zenoh.Config())
+
+
+# Test support: send data in background
+def send_data():
+    time.sleep(1)
+    for i in range(5):
+        session.put("key/expression", f"sample_{i}")
+
+
+threading.Thread(target=send_data, daemon=True).start()
+
+
+# [custom_channel]
+class CustomChannel:
+    def __init__(self, max_size=100):
+        self.samples = []
+        self.max_size = max_size
+        self.lock = threading.Lock()
+        self.condition = threading.Condition(self.lock)
+
+    def try_recv(self):
+        """Non-blocking receive"""
+        with self.lock:
+            return self.samples.pop(0) if self.samples else None
+
+    def recv(self):
+        """Blocking receive"""
+        with self.condition:
+            while not self.samples:
+                self.condition.wait()
+            return self.samples.pop(0)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        sample = self.recv()
+        if sample is None:
+            raise StopIteration
+        return sample
+
+    def add_sample(self, sample):
+        """Called by the callback to store samples"""
+        with self.condition:
+            self.samples.append(sample)
+            # Maintain max size
+            if len(self.samples) > self.max_size:
+                self.samples.pop(0)
+            # Notify one waiting thread that a sample is available
+            self.condition.notify()
+
+    def count(self):
+        """Return number of stored samples"""
+        with self.lock:
+            return len(self.samples)
+
+
+def create_custom_channel(
+    max_size: int = 100,
+) -> tuple[Callable[[zenoh.Sample], None], CustomChannel]:
+    """Factory function that returns (callback, handler) pair"""
+    channel = CustomChannel(max_size)
+
+    def on_sample(sample: zenoh.Sample) -> None:
+        # Store sample in the custom channel
+        channel.add_sample(sample)
+
+    return (on_sample, channel)
+
+
+# [custom_channel]
+
+count = 0
+# [custom_channel_usage]
+subscriber = session.declare_subscriber(
```

Comment:
I'm not sure that such a complex example of channel is pertinent when `queue.Queue` exists. 
A better example could be to something like this:
```python
class PrioritizedSamples:
    def __init__(self, maxsize: int):
        self.queue: queue.PriorityQueue[tuple[zenoh.Priority, zenoh.Sample]] = queue.PriorityQueue(maxsize)

    def put(self, sample: zenoh.Sample):
        self.queue.put((sample.priority, sample))

    def get(self) -> zenoh.Sample:
        return self.queue.get()[1]


session = zenoh.open()
channel = PrioritizedSamples(50)
subscriber = session.declare_subscriber(
    "key/expression", (channel.put, channel)
)

sample = subscriber.handler.get()
```

Reply:
I think one full custom channel example is enough. The goal is to demonstrate the internal mechanics of channel - the concept of pair (callback function, object), the full example works better for this I think